### PR TITLE
MP_bidirect_winograd with xdlops [FP32]

### PR DESCRIPTION
**Pull request by shurale-nkn**
Added new MP_bidirect_winograd solver with xdlops support.
This solver is tunable.

gfx908 time improvements ___without new solver tuning___

N | C | K | H | W | x | y | p | q | u | v | directioon | time reduce | develop | xdlops |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
  |   |   |   |   |   |   |   |   |   |   |   |   | best algo | this solver |  
256 | 128 | 128 | 28 | 28 | 3 | 3 | 1 | 1 | 1 | 1 | FWD | 15% | ConvMPBidirectWinograd<6-3> | ConvMPBidirectWinograd_xdlops<6-3>
256 | 128 | 128 | 28 | 28 | 3 | 3 | 1 | 1 | 1 | 1 | BWD | 15% | ConvMPBidirectWinograd<6-3> | ConvMPBidirectWinograd_xdlops<6-3>
256 | 256 | 256 | 14 | 14 | 3 | 3 | 1 | 1 | 1 | 1 | FWD | 39% | ConvBinWinogradRxSf2x3 | ConvMPBidirectWinograd_xdlops<5-3>
256 | 256 | 256 | 14 | 14 | 3 | 3 | 1 | 1 | 1 | 1 | BWD | 38% | ConvBinWinograd3x3U | ConvMPBidirectWinograd_xdlops<5-3>
256 | 512 | 512 | 7 | 7 | 3 | 3 | 1 | 1 | 1 | 1 | FWD | 54% | ConvMPBidirectWinograd<6-3> | ConvMPBidirectWinograd_xdlops<4-3>
256 | 512 | 512 | 7 | 7 | 3 | 3 | 1 | 1 | 1 | 1 | BWD | 51% | ConvMPBidirectWinograd<6-3> | ConvMPBidirectWinograd_xdlops<4-3>

### CODE REVIEW

**Comment by daniellowell** [comment]
@shurale-nkn Can you share target configs and performance data?

**Comment by shurale-nkn** [comment]
> @shurale-nkn Can you share target configs and performance data?

If you need more data, I can create a private thread.

**Comment by atamazov** [comment]
@shurale-nkn 
> If you need more data, I can create a private thread.

Better open an ticket for discussions at https://github.com/AMDComputeLibraries/MLOpen

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [inline src/solver/conv_MP_bidirectional_winograd.cpp:42]: Please remove surrounding parentheses in all #ifs

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [inline src/solver/conv_MP_bidirectional_winograd.cpp:53]: Do we really need to rename envvars? If so, then we have to update the documentation. Also all the users who use these envvars would need to update theirs scripts.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [inline src/solver/conv_MP_bidirectional_winograd.cpp:104]:
```suggestion
struct WinoOffsets {
    const size_t in, out, wei;
    WinoOffsets(size_t in_size, size_t wei_size, size_t out_size)
    : in(0), out(in_size), wei(in_size + out_size)
    {}
};
```
and so on. Make it const as much as possible.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [inline src/solver/conv_MP_bidirectional_winograd.cpp:399]: Why not MIOPEN_THROW?

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [inline src/solver/conv_MP_bidirectional_winograd.cpp:778]: I recommend putting this on hold until #203 is merged in. `RunAndMeasureSolution` will not be needed after that.

File: src/include/miopen/solver.hpp
Comment (atamazov) [inline src/include/miopen/solver.hpp:1559]: `<2,3>` is not needed.

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [inline src/solver/conv_MP_bidirectional_winograd.cpp:726]: Depends on `ConvHipImplicitGemmForwardV4R4Xdlops`?

**Review by atamazov — CHANGES_REQUESTED** [review]

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (shurale-nkn) [reply src/solver/conv_MP_bidirectional_winograd.cpp:104]: Make it const or make it struct???

File: src/include/miopen/solver.hpp
Comment (shurale-nkn) [reply src/include/miopen/solver.hpp:1559]: Do you know any other winograd2x3 wtih xdlops? Or are there other reasons?

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (shurale-nkn) [reply src/solver/conv_MP_bidirectional_winograd.cpp:726]: Yes

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (shurale-nkn) [reply src/solver/conv_MP_bidirectional_winograd.cpp:53]: Without renaming 1 envvar will manage 3 solvers
ConvMPBidirectWinograd
ConvMPBidirectWinograd_xdlops
ConvWinograd3x3MultipassWrW

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (shurale-nkn) [reply src/solver/conv_MP_bidirectional_winograd.cpp:399]: ``#if(MIOPEN_BACKEND_HIP && MIOPEN_USE_ROCBLAS)``
used only to avoid compiler error with OpenCl.
And ``MIOPEN_THROW`` used only for class functions to avoid errors if function must return an object.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (shurale-nkn) [reply src/solver/conv_MP_bidirectional_winograd.cpp]: Hmm, wait, it's too long... I'd better remove it .....right now.

File: src/include/miopen/solver.hpp
Comment (shurale-nkn) [reply src/include/miopen/solver.hpp:1412]: Needed! We don't have any other winograd f(2,3)  with xdlops

**Comment by asroy** [comment]
@shurale-nkn Has rocblas been tuned for these configs?

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (asroy) [reply src/solver/conv_MP_bidirectional_winograd.cpp:761]: We need to write a proper solver for it, ```ConvHipImplicitGemmForwardV4R4Xdlops``` is not using the best layout. I would need to know what layout you are using for the transformation kernels.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [reply src/solver/conv_MP_bidirectional_winograd.cpp]: Make it struct first. Then you can replace macro invocation with
```C++
const WinoOffsets winoOffset(in_size, wei_size, out_size);
...
// use:
... winoOffset.in ...
```

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [inline src/solver/conv_MP_bidirectional_winograd.cpp:404]:
```suggestion
                MIOPEN_THROW(miopenStatusBadParm, "ConvMPBidirectWinograd is not supported");
```
And this is internal error, I think.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [reply src/solver/conv_MP_bidirectional_winograd.cpp]: After #203, all Invokers must be able to measure elapsed time. Please double-check your Invoker and then press `[Resolve conversation]`. Thanks.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [inline src/solver/conv_MP_bidirectional_winograd.cpp:553]: Ditto.

File: src/include/miopen/solver.hpp
Comment (atamazov) [reply src/include/miopen/solver.hpp:1412]: But is wasn't instantiated. Now I see you've removed this declaration.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [reply src/solver/conv_MP_bidirectional_winograd.cpp:761]: It is necessary to clarify why this is necessary. Please add respective comment.

Please separately comment cases when:
- The `ConvHipImplicitGemmForwardV4R4Xdlops` solver is disabled (e.g. via environment).
- The iGEMM algo is totally disabled.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [reply src/solver/conv_MP_bidirectional_winograd.cpp:402]: Thanks for fixing.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [reply src/solver/conv_MP_bidirectional_winograd.cpp:53]: Formally speaking, no. The old names would look ambiguous without renaming, but this is acceptable cost of the backward compatibility. Thanks for fixing.

**Review by atamazov — CHANGES_REQUESTED** [review]
Minor issues

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (asroy) [reply src/solver/conv_MP_bidirectional_winograd.cpp:761]: Is this PR using Tensile or ```ConvHipImplicitGemmForwardV4R4Xdlops``` for the GEMM part?

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (shurale-nkn) [reply src/solver/conv_MP_bidirectional_winograd.cpp]: They were able to do this before #203. 
But now they can make tunning without 'RunAndMeasureSolution'.

But this solver during the tuning process, calls 'ConvHipImplicitGemmForwardV4R4Xdlops.Search()' and tunes ConvHipImplicitGemmForwardV4R4Xdlops solver without using its own kernels.

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (shurale-nkn) [reply src/solver/conv_MP_bidirectional_winograd.cpp:761]: @asroy 
>Is this PR using Tensile or ConvHipImplicitGemmForwardV4R4Xdlops for the GEMM part?

ConvHipImplicitGemmForwardV4R4Xdlops 


>We need to write a proper solver for it, ConvHipImplicitGemmForwardV4R4Xdlops is not using the best layout. I would need to know what layout you are using for the transformation kernels.

The original format presented in ConvMPBidirectWinograd is - GCNHW (for data and filter).
To work with ConvHipImplicitGemmForwardV4R4Xdlops the context is converted to NGCHW but N=1, G=G, C=C, H=1, W=H*G*W (for data), and in filter buffer used GNCHW (G=G, N=K, C=C, H=1, W=1). (The filter kernel code was changed specifically for this purpose)

Comment (shurale-nkn) [reply src/solver/conv_MP_bidirectional_winograd.cpp:761]: @atamazov 

>It is necessary to clarify why this is necessary. Please add respective comment.
>
>Please separately comment cases when:
>
>The ConvHipImplicitGemmForwardV4R4Xdlops solver is disabled (e.g. via environment).

As you can see, there is no such environment variable here
https://github.com/ROCmSoftwarePlatform/MIOpen/blob/3ff4d7fec01888ed18fad69565f703b82afee451/src/solver/conv_hip_implicit_gemm_v4r4_gen_xdlops.cpp#L439-L451

But if conv_hip_implicit_gemm_v4r4_gen_xdlops is disabled, this solver disabled to. (concept)

>The iGEMM algo is totally disabled.

Same principle as with GEMM.  This solver will be enabled.

I don't see any non-obvious points here. The previous solution worked on the same logic.

**Comment by shurale-nkn** [comment]
> @shurale-nkn Has rocblas been tuned for these configs?

@asroy 
If we speak about ConvMPBidirectWinograd algo in develop, then no, because ticket for tuning created a week ago. They wouldn't have had time to do it.

If about ConvMPBidirectWinograd_xdlops, than no, ConvHipImplicitGemmForwardV4R4Xdlops not tuned, used config from EuristicInit .

File: src/kernels/xform_bidirect_winograd_code.inc
Comment (Kirpich30000) [inline src/kernels/xform_bidirect_winograd_code.inc:94]: I suggest to name the variable by it's meaning. What it really do is switching layout. Please name it accordingly.

**Review by Kirpich30000 — CHANGES_REQUESTED** [review]
minor changes

**Review by Kirpich30000 — APPROVED** [review]
LGTM

**Comment by daniellowell** [comment]
@shurale-nkn Merge conflicts and blocking review. Please update this ASAP, otherwise it will not get into ROCm 3.10.

**Review by Kirpich30000 — APPROVED** [review]
LGTM

File: src/solver/conv_MP_bidirectional_winograd.cpp
Comment (atamazov) [reply src/solver/conv_MP_bidirectional_winograd.cpp:767]: OK

**Review by atamazov — APPROVED** [review]
Good job of reusing existing kernels/solvers
makeitso

**Comment by atamazov** [comment]
Please do not forget to document `MIOPEN_DEBUG_AMD_MP_BD_WINOGRAD_WORKSPACE_MAX` in separate PR. Also it would be nice to get rid of `// cppcheck-suppress unreadVariable` for `WinoOffsets`.

**Comment by shurale-nkn** [comment]
@daniellowell 
Update! Git doesn't display this, but the testing was completed successfully.

**Pull request merged**
